File: Data Structures/RainWaterTrapping.java

/* 
Problem Statement : Given n non-negative integers representing
an elevation map where the width of each bar is 1, 
compute how much water it can trap after raining.
LeetCode Question Number 42

Input: height = [0,1,0,2,1,0,1,3,2,1,2,1]
Output: 6
*/
import java.util.*;

public class RainWaterTrapping {

    // maximum water trapped
    public static int maxWater(int building[], int n) {

        int res = 0;

        for (int i = 1; i < n - 1; i++) {

            // Find maximum element on left of building
            int left = building[i];
            for (int j = 0; j < i; j++) {
                left = Math.max(left, building[j]);
            }

            // Find maximum element on right of building
            int right = building[i];
            for (int j = i + 1; j < n; j++) {
                right = Math.max(right, building[j]);
            }

            // Update maximum water value
            res += Math.min(left, right) - building[i];
        }
        return res;
    }

    // main
    public static void main(String[] args) {
        Scanner scn = new Scanner(System.in);
        int n;
        System.out.println("Enter length of the array");
        n = scn.nextInt();
        int building[] = new int[n];
        for (int i = 0; i < n; i++) {
            System.out.println("Enter elememts of array");
            building[i] = scn.nextInt();
        }

        int waterStored = maxWater(building, n);
        System.out.print("Water trapped due to rain is " + waterStored);
    }
}

File: Leetcode/Rainwater_trapping.cpp
//problem
//Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it is able to trap after raining.

//Examples
//Input: arr[]   = {3, 0, 2, 0, 4}
//Output: 7

//Input: arr[] = [0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]
//Output: 6

//The element of an array can store water only if there are higher bars on the left and right sides. 
//The amount of water to be stored in every element can be found out by finding the heights of bars on the left and right sides. 
//Now we will compute the amount of water that can be stored in every element of the array. 

//____________________________________________________________________________________________________________________________//
//Constraints:

//n == height.length
//0 <= n <= 3 * 104
//0 <= height[i] <= 105
//____________________________________________________________________________________________________________________________//
//Solution
#include <iostream>
using namespace std;
 
int calcWater(int arr[], int n)
{
    // initialize output
    int tot_water = 0;
 
    // maximum element on left and right
    int left_max = 0, right_max = 0;
 
    // indices to traverse the array
    int lo = 0, hi = n - 1;

    //start from first
    while (lo <= hi) {
        if (arr[lo] < arr[hi]) {
            if (arr[lo] > left_max)
                // update max in left
                left_max = arr[lo];
            else
                // water on curr element = max - curr
                tot_water+= left_max - arr[lo];
            lo++;
        }
        else {
            if (arr[hi] > right_max)
                // update right maximum
                right_max = arr[hi];
            else
                tot_water+= right_max - arr[hi];
            hi--;
        }
    }
 
    return  tot_water;
}
 
int main()
{
    int arr[] = { 0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1 };
    int n = sizeof(arr) / sizeof(arr[0]);
    cout << "Maximum water that can be accumulated is "
         << calcWater(arr, n);
}
// Time complexity o(n)
// Space complexity o(1)
